File: problems/树/226.翻转二叉树.cpp
#include <iostream>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* invertTree(TreeNode* root) {
    if (!root) return root;
    swap(root->left, root->right);
    invertTree(root->left);
    invertTree(root->right);
    return root;
}
void inorderPrint(TreeNode* root) {
    if (!root) return;
    inorderPrint(root->left);
    cout << root->val << " ";
    inorderPrint(root->right);
}
int main() {
    TreeNode* root = new TreeNode(4);
    root->left  = new TreeNode(2);
    root->right = new TreeNode(7);
    root->left->left  = new TreeNode(1);
    root->left->right = new TreeNode(3);
    root->right->left  = new TreeNode(6);
    root->right->right = new TreeNode(9);
    inorderPrint(invertTree(root)); cout << endl;  // 9 7 6 4 3 2 1
    inorderPrint(invertTree(new TreeNode(1))); cout << endl;  // 1
    cout << (invertTree(nullptr) == nullptr ? "null" : "?") << endl;  // null
    return 0;
}

File: problems/树/226.翻转二叉树.py
#!/usr/bin/env python3
"""
226. 翻转二叉树
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root: return root
    root.left, root.right = root.right, root.left
    invert_tree(root.left)
    invert_tree(root.right)
    return root

def inorder_list(root: Optional[TreeNode]):
    if not root: return []
    return inorder_list(root.left) + [root.val] + inorder_list(root.right)

if __name__ == "__main__":
    root1 = TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(7, TreeNode(6), TreeNode(9)))
    inverted = invert_tree(root1)
    print(inorder_list(inverted))  # [9, 7, 6, 4, 3, 2, 1]
    print(inorder_list(invert_tree(TreeNode(1))))  # [1]
    print(invert_tree(None))  # None
